### PR TITLE
Drop Py3.8 dependencies from `pyproject.toml`s and CI/CD workflows

### PR DESCRIPTION
**Pull request by stankudrow**
Partially addresses the issue #976 .
Should go after the PR #974 .

### CODE REVIEW

**Comment by stankudrow** [comment]
@dvarrazzo , problematic jobs:

- https://github.com/psycopg/psycopg/actions/runs/12470290784/job/34805159937?pr=977 - linux, c 3.9, min, pg 15 - OSError: Could not find library geos_c or load any of its variants ['libgeos_c.so.1', 'libgeos_c.so']

A striking solution: installing `libgeos-dev` -> https://stackoverflow.com/questions/19742406/could-not-find-library-geos-c-or-load-any-of-its-variants .

But why it failed, what have I missed?

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -90,6 +87,7 @@ jobs:
         if: ${{ matrix.impl == 'c' }}
         run: |
           echo "DEPS=$DEPS ./psycopg_c" >> $GITHUB_ENV
+          sudo apt-get install -y libgeos-dev
```
Comment (stankudrow) [inline .github/workflows/tests.yml:90]: Worked (but maybe it is a wrong place to have it here), but another error occurred: ModuleNotFoundError: No module named 'distutils.msvccompiler' ([link](https://github.com/psycopg/psycopg/actions/runs/12471615165/job/34808906125?pr=977))

**Comment by dvarrazzo** [comment]
The problem comes from the pinning of [shapely 1.7.0](https://github.com/stankudrow/psycopg/blob/4fd116dbd437e9381788d99b780269ded86b8f5a/tests/constraints.txt#L39C1-L39C8) in the `constraint.txt` file, which is the file we use to test that the code still works with a well defined set of old versions (and we don't accidentally introduce the requirement for a feature that is not available in a declared dependency).

Shapely 1.7.0 is [not available in a binary package](https://github.com/stankudrow/psycopg/blob/4fd116dbd437e9381788d99b780269ded86b8f5a/tests/constraints.txt#L39C1-L39C8) for Python 3.9. So we should bump that constraint to 1.8.0 for example, which [seems to support Python 3.9 and 3.10](https://pypi.org/project/shapely/1.8.0/#files).

**Comment by dvarrazzo** [comment]
For the same reason, in the same file, we should bump numpy to at least 1.20.0.

**Comment by stankudrow** [comment]
> For the same reason, in the same file, we should bump numpy to at least 1.20.0.

One moment...

Could you enlighten me, please, how have you found the problem so quickly and why these versions? I am interested in some rationale for better understanding.

File: tests/constraints.txt [hunk 2]
```diff
@@ -36,7 +36,7 @@ tomli == 2.0.1
 
 # Undeclared extras to "unblock" extra features
 
-shapely == 1.7.0
-# Warning: binary package only available up to python <= 3.8.
+shapely == 1.8.0
+# Warning: binary package only available up to python <= 3.9
```
Comment (stankudrow) [inline tests/constraints.txt:40]: Is this message correct?

Comment (dvarrazzo) [reply tests/constraints.txt:40]: TBH, the placement of that comment seems wrong, as the statement is valid both for numpy and for shapely, therefore I'd move it above the two.

The sentence is not precise because for example shapely 1.8 supports Python 3.10 too. However I think it's important that `3.9` is in the comment, because this allows to grep for it when it's time to remove support from a Python version.

Therefore I propose something like:

```
# Undeclared extras to "unblock" extra features
#
# Warning: the versions specified for these packages are the oldest versions
# offering a binary package for the oldest Python version we support.
#
# When the minimum supported Python version is increased, these dependencies
# might need to be updated.
#
# Grep help: the current minimum supported version is Python 3.9.
shapely == 1.8.0
numpy == 1.20.0
```

**Comment by dvarrazzo** [comment]
@stankudrow we have a documentation file with the list of chores needed when a new version of Python or Postgres is released: see https://github.com/psycopg/psycopg/blob/master/docs/release.rst

Could you maybe add a section "When dropping support to and old Python major version" and write down the operation we are doing now (updating this file, changing metadata, changing ci...)

**Comment by dvarrazzo** [comment]
There is actually a comment pointing at the need of updating the `constraint.txt` file:

https://github.com/psycopg/psycopg/blob/8868a72c8be121be2148f51a0beb545cd8ed332d/.github/workflows/tests.yml#L50-L54

You re-discovered it the hard way :smile:

**Comment by stankudrow** [comment]
> @stankudrow we have a documentation file with the list of chores needed when a new version of Python or Postgres is released: see https://github.com/psycopg/psycopg/blob/master/docs/release.rst
> 
> Could you maybe add a section "When dropping support to and old Python major version" and write down the operation we are doing now (updating this file, changing metadata, changing ci...)

Sure, will do.

**Comment by stankudrow** [comment]
> There is actually a comment pointing at the need of updating the `constraint.txt` file:
> 
> https://github.com/psycopg/psycopg/blob/8868a72c8be121be2148f51a0beb545cd8ed332d/.github/workflows/tests.yml#L50-L54
> 
> You re-discovered it the hard way 😄

Yeah, but I did not understand where to put `libgeos-dev` there 😅

File: tests/constraints.txt (hunk 2 above)
Comment (stankudrow) [reply tests/constraints.txt:40]: Ok, I'll just copypaste it.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (stankudrow) [reply .github/workflows/tests.yml:90]: @dvarrazzo , so, is it OK to leave here or there is a better place for it?

Comment (dvarrazzo) [reply .github/workflows/tests.yml:90]: No, we don't need it. The error only appeared because the wheel package for shapely 1.7.0 is not available for Python 3.9, therefore pip tried to install the requested version from source, and failed on the missing library.

Now that we are using a minimum version for which there is a wheel package for Python 3.9 we don't need this library.

Comment (stankudrow) [reply .github/workflows/tests.yml]: Gotcha, understood, removed. Thank you for your patience and explanations.

**Review by dvarrazzo — CHANGES_REQUESTED** [review]
Some of the CI entries dropped drop useful tests, so we need to move them to other versions, as suggested.

A quick grepping for `3\.8` finds work to do in:

- docs/src/install.rst
- compat.py modules and similar
- `build_macos_arm64.sh` (although this file is no more currently in use)
- typing hints here and there (Probably the tweaks added in  9ad4ea6d0bd6ae0e1a3394e03a385dee2198662d).

Grepping for `3\.9` finds:

- `psycopg_c/pyproject.toml`
- `psycopg/pyproject.toml` (`backports.zoneinfo` no more needed)

Grepping for `\b38\b` and `\b39\b` doesn't find much else (a `38` in  `tools/build/build_macos_arm64.sh` again).

Grepping for `3, 9` finds a few hits in compat modules and `tests/utils.py`

I don't have other places where to look at, at the moment.

Are these changes something you want to do in this MR or do you prefer to open a follow-up?

File: docs/release.rst [hunk 3]
```diff
@@ -32,7 +32,7 @@ How to make a psycopg release
   .. __: https://github.com/psycopg/psycopg/actions/workflows/packages-bin.yml
   .. __: https://github.com/psycopg/psycopg/actions/workflows/packages-pool.yml
 
-- Delete the ``wheelhouse`` directory there is one.
+- Delete the ``wheelhouse`` directory if there is one.
```
Comment (stankudrow) [inline docs/release.rst:35]: or this "if" is put wrong? Sorry for my English.

File: docs/release.rst [hunk 4]
```diff
@@ -110,3 +110,19 @@ When a new Python major version is released
...
+  in ``pyproject.toml`` files of the corresponding ``psycopg`` directories
+
+- Update necessary GitHub Actions workflow files in the ``.github/workflows/`` directory,
+  e.g., ``tests.yml`` and ``.3rd-party-tests.yml`` and so forth.
+
+- Bump needed versions in the ``tests/constraints.txt`` file.
+
+Examples:
+
+- `PR #977 <https://github.com/psycopg/psycopg/pull/977>`_
```
Comment (stankudrow) [inline docs/release.rst:128]: Looks good? Waiting for your suggestions, will be added ASAP.

File: docs/release.rst (hunk 3 above)
Comment (dvarrazzo) [reply docs/release.rst:35]: It's right, thank you, it was a typo :slightly_smiling_face:

**Comment by dvarrazzo** [comment]
Ah, just noticed: There are several cp38 in the `packages-bin.yml` workflow.

**Comment by stankudrow** [comment]
Just leave here -> running `python3 -m flake8 .` entails to the following errors:

```shell
╰─➤  python3 -m flake8 .
./tools/async_to_sync.py:244:15: E999 SyntaxError: invalid syntax
./tools/bump_version.py:104:15: E999 SyntaxError: invalid syntax
```

... because of `match-case` syntactic structure was introduced in Py3.10 ([StackOverflow](https://stackoverflow.com/questions/46701063/why-doesnt-python-have-switch-case-update-match-case-syntax-was-added-to-pyt)).

File: docs/release.rst [hunk 5]
```diff
@@ -110,3 +110,19 @@ When a new Python major version is released
 - Update the list of versions in ``tools/build/build_macos_arm64.sh`` to include
   the new version. Look for both the ``python_versions`` variable and the
   ``CIBW_BUILD`` environment variable.
+
+
+When dropping end-of-life Python versions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+- Update project metadata, ``requires-python`` and (maybe) package dependencies
+  in ``pyproject.toml`` files of the corresponding ``psycopg`` directories
```
Comment (dvarrazzo) [inline docs/release.rst:119]:
```suggestion
  in ``pyproject.toml`` files of the corresponding ``psycopg`` directories.
```

File: docs/release.rst [hunk 6]
```diff
@@ -110,3 +110,19 @@ When a new Python major version is released
...
+
+- Update necessary GitHub Actions workflow files in the ``.github/workflows/`` directory,
+  e.g., ``tests.yml`` and ``.3rd-party-tests.yml`` and so forth.
```
Comment (dvarrazzo) [inline docs/release.rst:122]:
```suggestion
  e.g., ``tests.yml``, ``3rd-party-tests.yml``, ``packages-bin.yml``.
```

File: docs/release.rst (hunk 4 above)
Comment (dvarrazzo) [reply docs/release.rst:128]: I would add (unless you want to do it in a different MR):

```
- Grep the source code for occurrences of the version to be dropped and the
  version following to find code branches depending on the version to remove.
  A few examples (dropping Python 3.8)::

     git grep -E -e '\b3\.8\b' -e '\b(cp)?38\b' -e '\b3, 8\b'
     git grep -E -e '\b3\.9\b' -e '\b(cp)?39\b' -e '\b3, 9\b'
```

**Comment by dvarrazzo** [comment]
> running python3 -m flake8 . entails to the following errors

I guess it depends on which version it's run. The CI runs it in Python 3.11. I personally run it with Python 3.10 (the stock version in ubuntu 22.04, which I'm still using). I don't see the need to keep compatibility of the development tools back to the same minimum version we want to make the library available.

File: psycopg/pyproject.toml [hunk 7]
```diff
@@ -13,8 +13,7 @@ classifiers = [
     "Operating System :: MacOS :: MacOS X",
     "Operating System :: Microsoft :: Windows",
     "Operating System :: POSIX",
-    "Programming Language :: Python :: 3",
-    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3 :: Only",
```
[the remark this answers is not in the document]
Comment (stankudrow) [reply psycopg/pyproject.toml:16]: https://pypi.org/classifiers/ - just claiming no more python 2 and since Py3 is the only supportable version, it may be fine unless Py4 is issued and both to be maintained.

I can remove this `Only` claim if you like it better.

File: tests/README.rst [hunk 8]
```diff
@@ -21,7 +21,7 @@ Test options
 
       $ pytest
       ========================= test session starts =========================
-      platform linux -- Python 3.8.5, pytest-6.0.2, py-1.10.0, pluggy-0.13.1
+      platform linux -- Python 3.9, pytest-6.0.2, py-1.10.0, pluggy-0.13.1
```
Comment (stankudrow) [inline tests/README.rst:24]: Should it be a micro part?

File: tests/acompat.py [hunk 9]
```diff
@@ -118,6 +118,7 @@ async def wait_timeout(self, timeout):
         await asyncio.wait_for(self.wait(), timeout)
 
 
+# TODO
```
Comment (stankudrow) [inline tests/acompat.py:121]: Today after some sleep.

File: tests/README.rst (hunk 8 above)
Comment (dvarrazzo) [reply tests/README.rst:24]: It's not important, it's just an illustrative example.

File: psycopg/pyproject.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (dvarrazzo) [reply psycopg/pyproject.toml:16]: I don't know the semantic difference between `3` and `3 :: Only`. I know that `3` doesn't mean `3.0`, so for me it had the same meaning as "not 2".

I am ok with this change, no problem. I just wanted to know if you had a better "official explanation".

**Comment by stankudrow** [comment]
> * `build_macos_arm64.sh` (although this file is no more currently in use)

So this `build_macos_arm64.sh` can be safely removed.

File: psycopg/pyproject.toml [hunk 10]
```diff
@@ -27,7 +26,7 @@ classifiers = [
     "Topic :: Software Development",
     "Topic :: Software Development :: Libraries :: Python Modules",
 ]
-requires-python = ">= 3.8"
+requires-python = ">= 3.9"
 dependencies = [
     "backports.zoneinfo >= 0.2.0; python_version < '3.9'",
```
Comment (pauloxnet) [inline psycopg/pyproject.toml:31]:
```suggestion
```
Can we remove `backports.zoneinfo` ?

**Comment by stankudrow** [comment]
Gosh, this should be recorded:
```shell
 diff --git a/tests/crdb/test_cursor.py b/tests/crdb/test_cursor.py
index 198feeb..aed5287 100644
--- a/tests/crdb/test_cursor.py
+++ b/tests/crdb/test_cursor.py
@@ -4,7 +4,7 @@
 from __future__ import annotations
 
 import json
-from queue import Queue
+from asyncio import Queue as AQueue
 from typing import Any
 from uuid import uuid4
 
ERROR sync and async files... out of sync!
```

Reference: https://github.com/psycopg/psycopg/actions/runs/12479753071/job/34829429651?pr=977

File: psycopg_pool/psycopg_pool/_acompat.py [hunk 11]
```diff
@@ -60,7 +42,7 @@ def current_task_name() -> str:
     return t.get_name() if t else "<no task>"
 
 
-class Queue(_GQueue[T]):
+class Queue(queue.Queue[T]):
```
Comment (stankudrow) [inline psycopg_pool/psycopg_pool/_acompat.py:45]: Leave it like this or remove completely?

File: docs/release.rst (hunk 4 above)
Comment (stankudrow) [reply docs/release.rst]: Changed, is it still good?

**Comment by stankudrow** [comment]
@dvarrazzo , finally linting stages passes.

I noticed that you suggested some commits and I didn't accept them because I applied them during other changes. But I don't want you to steal your coauthorship, so, if you don't mind, feel free to suggest some commits, I'll accept them.

**Comment by stankudrow** [comment]
Darn, MacOS fails: https://github.com/psycopg/psycopg/actions/runs/12480024372/job/34830158538?pr=977 .

File: docs/release.rst [hunk 12]
```diff
@@ -110,3 +110,27 @@ When a new Python major version is released
...
+  However, favouring smaller pull requests is convenient and easy to review.
+  An example for grepping `end-of-life <https://endoflife.date/python>` Python 3.8::
+
+     git grep -E -e '\b3\.8\b' -e '\b(cp)?38\b' -e '\b3, 8\b'
+
+- Consider proceeding steps described in the ``.github/workflows/lint.yml`` before ``git push``.
+
+Examples:
+
+- `PR #977 <https://github.com/psycopg/psycopg/pull/977>`_
```
Comment (stankudrow) [inline docs/release.rst:136]: @dvarrazzo , @pauloxnet , let's have another round.

File: tests/acompat.py [hunk 13]
```diff
@@ -118,17 +118,5 @@ async def wait_timeout(self, timeout):
...
-        # Always specify a timeout to make the wait interruptible.
-        if timeout is None:
-            timeout = 24.0 * 60.0 * 60.0
-        return super().get(block=block, timeout=timeout)
-
-
-class AQueue(asyncio.Queue):  # type: ignore[type-arg]
-    pass
+Queue = queue.Queue
+AQueue = asyncio.Queue
```
Comment (stankudrow) [inline tests/acompat.py:122]: Should it be got `from psycopg.psycopg._acompat import AQueue, Queue` or just leaving it so?

**Comment by stankudrow** [comment]
https://github.com/psycopg/psycopg/actions/runs/12486112485/job/34845795320?pr=977

```shell
Run docker pull postgres:12, libpq: master
"docker pull" requires exactly 1 argument.
See 'docker pull --help'.

Usage:  docker pull [OPTIONS] NAME[:TAG|@DIGEST]

Download an image from a registry
Error: Process completed with exit code 1.
```

**Comment by stankudrow** [comment]
I need help...

File: tools/build/build_macos_arm64.sh [hunk 14]
```diff
@@ -9,7 +9,7 @@
 
 set -euo pipefail
 
-python_versions="3.8.19 3.9.19 3.10.14 3.11.9 3.12.5 3.13.0"
+python_versions="3.9.19 3.10.14 3.11.9 3.12.5 3.13.0"
```
Comment (stankudrow) [inline tools/build/build_macos_arm64.sh:12]: This file to be deleted in another PR because its deletion has little to do with dropping Python 3.8.

**Comment by stankudrow** [comment]
Totals:

- importing some entities from `collections.abc` in lieu of `typing` module -> to be investigated
- replacing existing equivalents of entities from standard library (like AQueue which is just asyncio.Queue, or asleep which is asyncio.sleep under the hood) -> to be removed/reconsidered in other PRs
- removing `tools/build/build_macos_arm64.sh` - in another PR (this is fine)

@dvarrazzo , @pauloxnet , let's finalise this PR and merge its trade-off version.

**Review by pauloxnet — APPROVED** [review]
LGTM

File: psycopg_pool/psycopg_pool/abc.py [hunk 15]
```diff
@@ -28,7 +27,6 @@
 
 # Callbacks to pass the pool to on connection failure
 ConnectFailedCB: TypeAlias = Callable[["ConnectionPool[Any]"], None]
-AsyncConnectFailedCB: TypeAlias = Union[
-    Callable[["AsyncConnectionPool[Any]"], None],
-    Callable[["AsyncConnectionPool[Any]"], Awaitable[None]],
+AsyncConnectFailedCB: TypeAlias = Callable[
+    ["AsyncConnectionPool[Any]"], Optional[Awaitable[None]]
```
Comment (dvarrazzo) [inline psycopg_pool/psycopg_pool/abc.py:31]: I don't think that this definition is better than the previous one. I do understand that my comment that `Union` could be dropped in Python 3.8 was wrong (`Callable | Callable` is only supported from Python 3.10), but I would rather keep an `Union` here.

**Review by dvarrazzo — CHANGES_REQUESTED** [review]
We are pretty close! Just a few adjustments to make, but I'm ok to not add (or remove) anything else to the MR.

File: docs/release.rst [hunk 16]
```diff
@@ -110,3 +110,27 @@ When a new Python major version is released
...
+
+- Bump versions in the ``tests/constraints.txt`` file if it is necessary.
+
+- You may grep throughout the project for occurrences of a version to be dropped.
+  However, favouring smaller pull requests is convenient and easy to review.
+  An example for grepping `end-of-life <https://endoflife.date/python>` Python 3.8::
+
+     git grep -E -e '\b3\.8\b' -e '\b(cp)?38\b' -e '\b3, 8\b'
+
+- Consider proceeding steps described in the ``.github/workflows/lint.yml`` before ``git push``.
```
[the remark this answers is not in the document]
Comment (stankudrow) [reply docs/release.rst:132]: Ok, I remove it from release docs.

File: psycopg_pool/psycopg_pool/abc.py (hunk 15 above)
Comment (stankudrow) [reply psycopg_pool/psycopg_pool/abc.py:31]: Will do...

File: docs/basic/install.rst [hunk 17]
```diff
@@ -22,7 +22,7 @@ Supported systems
 
 The Psycopg version documented here has *official and tested* support for:
 
-- Python: from version 3.8 to 3.13
+- Python: from version 3.9 to 3.13
 
   - Python 3.6 supported before Psycopg 3.1
   - Python 3.7 supported before Psycopg 3.2
```
[the remark this answers is not in the document]
Comment (stankudrow) [reply docs/basic/install.rst:28]: Added and reversed ...

File: .github/workflows/tests.yml [hunk 18]
```diff
@@ -26,32 +26,29 @@ jobs:
       matrix:
         include:
           # Test different combinations of Python, Postgres, libpq.
-          - {impl: python, python: "3.8", postgres: "postgres:17", libpq: oldest}
           - {impl: python, python: "3.9", postgres: "postgres:16", libpq: master}
           - {impl: python, python: "3.10", postgres: "postgres:15"}
           - {impl: python, python: "3.11", postgres: "postgres:14"}
           - {impl: python, python: "3.12", postgres: "postgres:13", libpq: newest}
           - {impl: python, python: "3.13", postgres: "postgres:12"}
```
[the remark this answers is not in the document]
Comment (stankudrow) [reply .github/workflows/tests.yml:33]: I hope I wrote it right.

File: .github/workflows/tests.yml [hunk 19]
```diff
@@ -26,32 +26,31 @@ jobs:
...
           # Test different combinations of Python, Postgres, libpq.
-          - {impl: python, python: "3.8", postgres: "postgres:17", libpq: oldest}
+          - {impl: python, python: "3.9", postgres: "postgres:17", libpq: oldest}
           - {impl: python, python: "3.9", postgres: "postgres:16", libpq: master}
-          - {impl: python, python: "3.10", postgres: "postgres:15"}
-          - {impl: python, python: "3.11", postgres: "postgres:14"}
-          - {impl: python, python: "3.12", postgres: "postgres:13", libpq: newest}
+          - {impl: python, python: "3.10", postgres: "postgres:16"}
+          - {impl: python, python: "3.11", postgres: "postgres:15"}
+          - {impl: python, python: "3.12", postgres: "postgres:14", libpq: newest}
```
[the remark this answers is not in the document]
Comment (stankudrow) [reply .github/workflows/tests.yml:32]: Done, leaving [this comment](https://github.com/psycopg/psycopg/pull/977/files#r1896102978) unresolved for cross-validation sake.

**Comment by stankudrow** [comment]
Daniele and Paolo, thank you for your sharp attention, enormous patience and massive help.

**Review by dvarrazzo — APPROVED** [review]

**Review by dvarrazzo — APPROVED** [review]

**Comment by dvarrazzo** [comment]
Merged! Thank you very much for your patience, @stankudrow!

**Comment by stankudrow** [comment]
> Merged! Thank you very much for your patience, @stankudrow!

Thank you.
